decisions/0004-remixing-react-router.md

```diff
@@ -0,0 +1,173 @@
+
+```js
+function shouldRevalidate({ defaultShouldRevalidate }) {
+  // Don't revalidate for this case
+  if (someEdgeCase()) {
+    return false;
+  }
+
+  // Otherwise, do what we would have done by default
+  return defaultShouldRevalidate;
+}
+```
+
+### `<ScrollRestoration getKey>` prop
+
+In Remix, the `<ScrollRestoration>` component made an assumption that we would always restore scroll position based on `location.key`. If the key was the same as a prior location we knew the scroll position for, then we knew you had been there before and we should restore. This works great for back/forward navigations but it's a bit overly restrictive. You cannot choose tgo restore scroll based on anything other than `key`. Twitter has a great implementation of this as you click around in their left nav bar - your tweet feed is always at the same place when you click back to it - even though its a _new_ location in the history stack. This is because they're restoring by pathname here instead of `location.key`. Or maybe you want to maintain scroll position for all routes under a given pathname and you thus want to use a portion of the pathname as the scroll restoration key.
+
+In React Router we now accept an optional `<ScrollRestoration getKey>` prop where you provide a function that returns the key to use for scroll restoration:
+
+```js
+function getKey(location: Location, matches: DataRouteMatch[]) {
+  // Restore by pathname on /tweets
+  if (location.pathname === "/tweets") {
+    return location.pathname;
+  }
+  // Otherwise use the key
+  return location.key;
+}
+```
+
+**Backwards Compatibility**
+
+We're ok here since the new prop is optional and defaults to using `location.key`
+
+### `<Link resetScroll="false">` prop
+
+In addition to `<ScrollRestoration>` handling "restoring" scroll position on previously visited routes. It also handles "resetting" scroll position back to the top on _new_ routes. This is not always desireable if you're clicking around inside a tabbed view or something, so we've introduced a new `<Link resetScroll>` prop that lets you disable the scroll reset behavior _for a given navigation_. Note that this "resetting" logic happens if and only if we cannot restore scroll to a previously known location for that scroll restoration key.
+
+### `useRevalidator()` hook
+
+This has been a long time coming - see https://github.com/remix-run/remix/discussions/1996 🙂
+
+### No distinction between Error and Catch boundaries
+
+TODO: Ryan to fill in some more details here.
+
+The differentiation between error and catch prove to be a bit vague over time and a source of confusion for developers. We chose to go with just a single `errorElement` in the router for simplicity. If you throw anything, it ends up in the error boundary and propagates accordingly.
+
+**Backwards Compatibility**
+
+We will handle this in Remix where it's `errorElement` will decide if it should render via the `ErrorBoundary` or `CatchBoundary`.
```

Comment:
```suggestion
We will handle this in Remix where its `errorElement` will decide if it should render via the `ErrorBoundary` or `CatchBoundary`.
```

Reply:
About this part, I understand that things will not change in Remix, and we'll keep using `ErrorBoundary` and `CatchBoundary`.

If that's indeed the case, why aren't we clearing up the confusion in Remix too by having a single function for both?

Reply:
yeah - still some stuff up in the air there.  We definitely plan to drop to a single function in v2, I think initially we were thinking we'd just drop `CatchBoundary` and handle everything in `ErrorBoundary` in v2.  However, I think what you're proposing is introduce a net new (maybe `ErrorElement`) in Remix v1 and deprecating both `ErrorBoundary`/`CatchBoundary` in v1 (and dropping in v2).  I kind of like that idea - I'll run that by the team as well!